src/Parsers/ExpressionElementParsers.cpp

```diff
@@ -834,10 +826,12 @@ bool ParserNumber::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
 
     auto try_read_float = [&](const char * it, const char * end)
     {
+        std::string buf(it, end); /// Copying is needed to ensure the string is 0-terminated.
```

Comment:
It seems we are already copying the data into buf. Why isn't that enough?

Reply:
Where do we copy it? I missed probably.

Reply:
Not always:
https://github.com/ClickHouse/ClickHouse/blob/eea5a2c521e8143d1b90381efeaddb7da460caf9/src/Parsers/ExpressionElementParsers.cpp#L856-L860

Reply:
As a suggestion, we can create a new string in this case separately, it looks like all other cases are always null-terminated.
```
    /// NaN and Inf
    if (pos->type == TokenType::BareWord)
    {
        std::string buf(it, end);
        return try_read_float(buf.c_str(), buf.c_str() + buf.size());
    }
```